ocaml/typing/printtyp.ml

```diff
@@ -2065,17 +2065,68 @@ let expand_module_type =
   ref ((fun _env _mty -> assert false) :
       Env.t -> module_type -> module_type)
 
-let rec tree_of_modtype ?(ellipsis=false) = function
+(** How to abbreviate signatures *)
+module Abbrev = struct
+  (* The code is substantially simpler if [width] is mutable. Strictly speaking, [depth]
+     doesn't have to be mutable here but mixed mutability would be quite confusing. *)
+  type t =
+    { (* To what depth to unfold the module tree *)
+      mutable depth : int
+      (* How many signature items to print in total across all signatures *)
+    ; mutable width : int
+    }
+
+  (** Standard abbreviation heuristic *)
+  let abbrev () =
+    { depth = 4
+    ; width = 16
+    }
+
+  (** Don't print any signature items *)
+  let ellipsis () =
+    { depth = 0
+    ; width =  0
+    }
+
+  (** Should we print anything in this signature *)
+  let exhausted = function
+    | Some {depth; width} -> depth <= 0 || width <= 0
+    | None -> false
+
+  (** Run [f] at one deeper unfolding level *)
+  let deeper t f =
+    match t with
+    | Some t ->
+        let saved = t.depth in
+        t.depth <- t.depth - 1;
+        let x = f () in
+        t.depth <- saved;
+        x
+    | None -> f ()
+
+  (** Reduce the remaining width by the number of items in [sg] and return the number of
+      items to print in [sg] and a flag that inidicates whether [sg] is being trimmed. *)
+  let items t sg =
+    match t with
+    | Some t ->
+        let n = List.length sg in
+        let k = min t.width n in
+        t.width <- t.width - n;
+        Some k, (k < n)
+    | None ->
+        None, false
+end
+
+let rec tree_of_modtype ?abbrev = function
```

Comment:
It just comes down to taste, but I tend to lean away from optional parameters, unless the omitted form is much more common than the explicit form -- not the case here.

Reply:
There are quite a few calls further down in the module which would have to be changed. I lean towards making diffs as small as possible to make upstream merging easier but happy to change this if you feel more strongly about this than I do.

Reply:
No, I don't feel strongly.